coders/heic.c

```diff
@@ -822,128 +985,97 @@ static MagickBooleanType WriteHEICImage(const ImageInfo *image_info,
   heif_context=heif_context_alloc();
   heif_image=(struct heif_image*) NULL;
   heif_encoder=(struct heif_encoder*) NULL;
+
   do
   {
-    const Quantum
-      *p;
-
 #if LIBHEIF_NUMERIC_VERSION >= 0x01040000
     const StringInfo
       *profile;
 #endif
 
-    int
-      stride_y,
-      stride_cb,
-      stride_cr;
-
     struct heif_error
       error;
 
     struct heif_writer
       writer;
 
-    uint8_t
-      *p_y,
-      *p_cb,
-      *p_cr;
+    enum heif_colorspace 
+      colorspace;
+
+    enum heif_chroma 
+      chroma;
+
+    MagickBooleanType 
+      isAvif;
+
+
+    colorspace=heif_colorspace_YCbCr;
+    chroma=heif_chroma_420;
+    isAvif=MagickFalse;
 
     /*
       Get encoder for the specified format.
     */
 #if LIBHEIF_NUMERIC_VERSION > 0x01060200
     if (LocaleCompare(image_info->magick,"AVIF") == 0)
-      error=heif_context_get_encoder_for_format(heif_context,
-        heif_compression_AV1,&heif_encoder);
+      {
+        error=heif_context_get_encoder_for_format(heif_context,
+          heif_compression_AV1,&heif_encoder);
+        isAvif=MagickTrue;
+      }
     else
 #endif
       error=heif_context_get_encoder_for_format(heif_context,
         heif_compression_HEVC,&heif_encoder);
     status=IsHeifSuccess(&error,image,exception);
     if (status == MagickFalse)
       break;
-    /*
-      Transform colorspace to YCbCr.
-    */
-    if (image->colorspace != YCbCrColorspace)
-      status=TransformImageColorspace(image,YCbCrColorspace,exception);
-    if (status == MagickFalse)
-      break;
+
+    printf("Colorspace: %d\n", image->colorspace);
+    if (IssRGBCompatibleColorspace(image->colorspace) && isAvif)
+      {
+        printf("sRGB colorspace!\n");
+        colorspace=heif_colorspace_RGB;
+        chroma=IsImageOpaque(image, exception) ? 
+          heif_chroma_interleaved_RGB : 
+          heif_chroma_interleaved_RGBA;
+      } 
+    else if (image->colorspace != YCbCrColorspace)
+      {
+        /*
+        Transform colorspace to YCbCr.
+        */
+        status=TransformImageColorspace(image,YCbCrColorspace,exception);
+        if (status == MagickFalse)
+          break;
```

Comment:
Looks like an old bug in the code we should exit here when this fails.

Reply:
Sorry, do you mean `return status`? It looks like there is some cleanup going on at the end of the function.

Reply:
Nevermind.. I read this wrong. Did not see we where inside the `do while`.